integration-tests/src/test_loop/utils/transactions.rs

```diff
@@ -201,3 +206,92 @@ pub fn get_node_data<'a>(node_datas: &'a [TestData], account_id: &AccountId) ->
     }
     panic!("RPC client not found");
 }
+
+/// Run a transaction until completion and assert that the result is "success".
+/// Returns the transaction result.
+pub fn run_tx(
+    test_loop: &mut TestLoopV2,
+    tx: SignedTransaction,
+    node_datas: &[TestData],
+    maximum_duration: Duration,
+) -> Vec<u8> {
+    let tx_res = execute_tx(test_loop, tx, node_datas, maximum_duration).unwrap();
+    assert_matches!(tx_res.status, FinalExecutionStatus::SuccessValue(_));
+    match tx_res.status {
+        FinalExecutionStatus::SuccessValue(res) => res,
+        _ => unreachable!(),
+    }
+}
+
+/// Submit a transaction and wait for the execution result.
+/// For invalid transactions returns an error.
+/// For valid transactions returns the execution result (which could have an execution error inside, check it!).
+pub fn execute_tx(
+    test_loop: &mut TestLoopV2,
+    tx: SignedTransaction,
+    node_datas: &[TestData],
+    maximum_duration: Duration,
+) -> Result<FinalExecutionOutcomeView, InvalidTxError> {
+    // Last node is usually the rpc node
+    let rpc_node_id = node_datas.len().checked_sub(1).unwrap();
```

Comment:
I was wondering does it matter to which node we send the request to? Why would we want to send to RPC node, and is it true that the last node is the RPC node?

Reply:
> I was wondering does it matter to which node we send the request to? Why would we want to send to RPC node

I wanted to use an RPC node because it tracks all shards. I'm not sure if I can submit a transaction to a node that doesn't track a shard. There's some forwarding logic, but I haven't read into it. Apart from that I need to read the execution outcome, and I think that won't be available on a node that doesn't track the shard.

> and is it true that the last node is the RPC node?

AFAIU the current logic of assigning node roles is that the first few nodes are producers and validators, which track a single shard, ant the rest of them are RPC nodes which track all shards. This means that the last node would be an RPC node.

https://github.com/near/nearcore/blob/9932e95c5c49262189bc4737366ce46bb2b953c8/integration-tests/src/test_loop/builder.rs#L185-L199

(added in https://github.com/near/nearcore/pull/11696 by @wacban)

It's a bit implicit, it would be nicer to have a builder method that specifies which nodes should be the rpc nodes.

Reply:
Alright! I would have been slightly happier if we had some check in place to confirm whether we are dealing with a node that tracks all shards or not. If this function is used unwisely, it could fail without there being an actual error in logic.

Reply:
Yeah I think the RPC nodes need some kind of rework. I think ideally you could specify in the builder that some of the clients will be RPCs and then testloop utilities could automatically pick a known RPC node and perform operations on it.
For now I implemented something that works with the existing code.